### DISCUSSION TITLE
Learning asyncio module

### DISCUSSION THREAD

**Discussion by ebolisa** [micropython/micropython#11014] [General]
Hi,

I'm trying to do multitasking using the asyncio module with the following code on a pico w:
```
import uasyncio
import gc
gc.collect()

async def clock():
    while True:
        t = utime.localtime()
        print("Clock: {:02d}:{:02d}:{:02d}".format(t[3], t[4], t[5]))
        await uasyncio.sleep(1)
    
    
async def temperatures():
    while True:
        gc.collect()
        t = utime.localtime()
        print("Temps: {:02d}:{:02d}:{:02d}".format(t[3], t[4], t[5]))
        await uasyncio.sleep(6)

    
async def main():
    uasyncio.create_task(clock())
    uasyncio.create_task(temperatures())
    while True:
        await clock()
        await temperatures()
    
uasyncio.run(main())
```
and these are the results:
```
Clock: 11:41:56
Clock: 11:41:56
Temps: 11:41:57
Clock: 11:41:57
Clock: 11:41:57
Clock: 11:41:58
Clock: 11:41:58
Clock: 11:41:59
Clock: 11:41:59
Clock: 11:42:00
Clock: 11:42:00
Clock: 11:42:01
Clock: 11:42:01
Clock: 11:42:02
Clock: 11:42:02
Temps: 11:42:03
Clock: 11:42:03
Clock: 11:42:03
```
Apparently, the temperature routine is called every 6 seconds (approx.) but the clock is called twice/second.
Why is that?

TIA

**Comment by peterhinch** [comment]
You have created two instances of the `clock` task. `create_task` creates one, then you `await clock()` which creates another. The line `await temperatures()` will never execute because `clock` (and hence its `await`) never terminates. Try this, which will create one instance of each:
```python
async def main():
    uasyncio.create_task(clock())
    while True:
        await temperatures()
```
You might like to look at [the uasyncio tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md).

**Reply by ebolisa** [reply]
Thank you Peter.
I was following the doc [here](https://docs.micropython.org/en/latest/library/uasyncio.html#module-uasyncio). Obviously, I didn't understand the process.

**Comment by peterhinch** [comment]
When the following executes
```python
async def foo():
    await bar()
    do_something()
```
the `do_something()` line will only run when `bar()` terminates. By contrast in
```python
async def foo():
    asyncio.create_task(bar())
    do_something()
```
`bar()` runs concurrently. So long as `bar()` is correctly written as asynchronous code, `do_something()` will run immediately.

The `main()` routine is commonly designed never to terminate because if it does, `uasyncio` terminates and you get back to the REPL.

**Reply by ebolisa** [reply]
Got it, thanks again.

I had an alarm clock fully controlled/configured by web and running with threads and timers but, it wasn't stable.
This passed weekend, I removed all the timers the threads routines and added the uasyncio module. So far, the clock has been running flawlessly.

**Reply by peterhinch** [reply]
In general my advice is to use `uasyncio` in preference to timers, and definitely in preference to threads. Clearly there is a "where possible" caveat. Where you do use timers or threads, it's wise to be clear on exactly what they offer that `uasyncio` does not.
